### PR TITLE
Add PartialEq and Eq to Mutex

### PR DESCRIPTION
**Pull request by Luro02**
This will allow to 2 `Mutex`es to be compared with each other.

I think, that there shouldn't be an `unwrap` and instead an `expect`, but I don't know what the message should be (the string, that is passed to `expect`).

Should I change the trait bounds to `<T: ?Sized + PartialEq>` and would it make sense to also implement `PartialEq` and `Eq` for [`MutexGuard`](https://doc.rust-lang.org/std/sync/struct.MutexGuard.html)? (maybe I should also implement the other missing traits, `Hash`, `PartialOrd`, `Ord`?)

### CODE REVIEW

**Comment by sfackler** [comment]
It seems pretty scary to take locks out in `==` - it'd be too easy for user code to deadlock accidentally imo.

**Comment by Luro02** [comment]
Would `try_lock` be better?

**Comment by sfackler** [comment]
A definition of equality that changed nondeterministically if any other thread held the lock at the time doesn't seem like an ideal behavior to me.

**Comment by Luro02** [comment]
So should this implementation make a shared reference from the internal unsafe cell and then compare those refs and ignore the lock?

**Comment by sfackler** [comment]
No, that is profoundly unsound for the same reason covered in #67229

**Comment by hellow554** [comment]
I personally doubt that is useful.
Can you provide a real world example?

**Comment by lukaslueg** [comment]
IMHO this is spooky action at a distance. Comparing values behind a `Mutex` will look like to just work, while in fact there is a hidden lock-action on every single comparison. I don't think the ergonomic gains are not worth hiding the fact that there is a lock involved.

Also, I think the pointer comparison is actually not correct: A separate thread can have the lock taken and change the value behind the lock. The line `A == A` is `true` in this case, while actually it is not defined as long as you don't hold the lock!

File: src/libstd/sync/mutex.rs [hunk 1]
```diff
@@ -463,6 +463,20 @@ impl<T: ?Sized + fmt::Display> fmt::Display for MutexGuard<'_, T> {
     }
 }
 
+#[unstable(feature = "mutex_partial_eq", issue = "0")]
+impl<T: PartialEq> PartialEq for Mutex<T> {
+    fn eq(&self, other: &Self) -> bool {
+        if ptr::eq(&self, &other) {
+            true
```
Comment (cuviper) [inline src/libstd/sync/mutex.rs]: As mentioned [on the forum](https://users.rust-lang.org/t/is-it-possible-to-compare-the-same-unsafecell-with-itself/35626/4?u=cuviper), reflexive equality can only be assumed for `T: Eq`. You can see [`ArcEqIdent`](https://github.com/rust-lang/rust/blob/7de9402b77ded0d8ec9e1c554521b2121449ef2b/src/liballoc/sync.rs#L1750-L1754) and [`RcEqIdent`](https://github.com/rust-lang/rust/blob/7de9402b77ded0d8ec9e1c554521b2121449ef2b/src/liballoc/rc.rs#L1185-L1189) for the way to specialize this though.

Comment (Luro02) [reply src/libstd/sync/mutex.rs]: Would it work to simply compare the value to itself with
```rust
let reference = self.lock().unwrap();
reference.eq(& reference)
```

**Comment by Luro02** [comment]
@hellow554 My motivation for this PR is the following:

I have written a crate called [`sub_cursor`](https://github.com/Luro02/sub_cursor), which allows to have something similar to slices, but on readers and writers.
For example, I have the following cursor
```rust
Cursor::new(vec![1, 2, 3, 4, 5, 6, 7, 8, 9])
```
I can make several sub-cursor, by wrapping the above cursor in a sub-cursor
```rust
let sub_cursor = SubCursor::builder(Cursor::new(vec![1, 2, 3, 4, 5, 6, 7, 8, 9])).build().unwrap();
```

Now lets say I need a cursor, which has only access to the first 3 items of the cursor, I can do this by calling `fn sub_cursor(&self, start: usize, end: usize) -> SubCursor<T>`.
```rust
let sub_cursor_2 = sub_cursor.sub_cursor(0, 3);

let mut result = vec![];
sub_cursor_2.read_to_end(&mut result)?;

assert_eq!(vec![1, 2, 3], result);
```
Internally an `Arc` is used to allow multiple SubCursors, to have access to the same underlying reader or writer. I wanted to make this struct thread-safe, so I used an `Arc` and a `Mutex`, because `RefCell` isn't thread-safe and the definition of the traits `Read`, `Write` and `Seek` all require mutable access to the underlying cursor, so it's not enough to just wrap it in an `Arc`.

This library could be useful in for example a file format parser, where there are multiple files embedded into the file, like for example a zip file. You don't have to read the file into memory and can give access to the file by for example such a function:
```rust
fn get_file(&mut self, name: &str) -> impl Read + Seek + Clone;
```
(btw I know, that the above will most certainly not work, because impl doesn't allow, to return 
  a type with multiple traits?)

My solution works, but it's missing a `PartialEq` implementation, so I can't conveniently assert inside tests, if 2 `SubCursor`s are the same.

After a bit of googling I stumbled upon rust-lang/rfcs#2055, but there wasn't any progress made since 2017, so I decided to take it in my own hands, to make it possible to check 2 `Mutex` for equality and this is the result.

-----
To me it seems like it's impossible to compare several mutex in a way, without locking the mutex and like I was already told, it's very unsafe to compare the internal `UnsafeCell`s.
Therefore I think this PR should be closed, but maybe someone has a solution for this problem.

File: src/libstd/sync/mutex.rs (hunk 1 above)
Comment (cuviper) [reply src/libstd/sync/mutex.rs]: I think so, relying on auto-`Deref` through the `MutexGuard` to the inner type. Or you could make that explicit with `*reference == *reference`. (Personally, I'd call that local `guard`.)

**Comment by steveklabnik** [comment]
Someone asked about this [on reddit](https://www.reddit.com/r/rust/comments/e86t53/hey_rustaceans_got_an_easy_question_ask_here/fah3kuw/?context=3) and my opinion remains the same, for what it's worth: this is too much magic.

**Pull request closed without merging**